js/codemeta_generation.js

```diff
@@ -115,49 +128,65 @@ function generatePersons(prefix) {
-            "@type": "SoftwareSourceCode",
-        };
-
-        let licenses = getLicenses();
-        if (licenses.length > 0) {
-            doc["license"] = (licenses.length === 1) ? licenses[0] : licenses;
-        }
+function buildDoc() {
+    var doc = {
+        "@context": CODEMETA_CONTEXT_URL,
+        "@type": "SoftwareSourceCode",
+    };
 
-        // Generate most fields
-        directCodemetaFields.forEach(function (item, index) {
-            doc[item] = getIfSet('#' + item)
-        });
+    let licenses = getLicenses();
+    if (licenses.length > 0) {
+        doc["license"] = (licenses.length === 1) ? licenses[0] : licenses;
+    }
 
-        doc["funder"] = generateShortOrg('#funder', doc["affiliation"]);
+    // Generate most fields
+    directCodemetaFields.forEach(function (item, index) {
+        doc[item] = getIfSet('#' + item)
+    });
 
-        // Generate simple fields parsed simply by splitting
-        splittedCodemetaFields.forEach(function (item, index) {
-            const id = item[0];
-            const separator = item[1];
-            const value = getIfSet('#' + id);
-            if (value !== undefined) {
-                doc[id] = value.split(separator).map(trimSpaces);
-            }
-        });
+    doc["funder"] = generateShortOrg('#funder', doc["affiliation"]);
 
-        // Generate dynamic fields
-        var authors = generatePersons('author');
-        if (authors.length > 0) {
-            doc["author"] = authors;
-        }
-        var contributors = generatePersons('contributor');
-        if (contributors.length > 0) {
-            doc["contributor"] = contributors;
+    // Generate simple fields parsed simply by splitting
+    splittedCodemetaFields.forEach(function (item, index) {
+        const id = item[0];
+        const separator = item[1];
+        const value = getIfSet('#' + id);
+        if (value !== undefined) {
+            doc[id] = value.split(separator).map(trimSpaces);
         }
+    });
 
-        codemetaText = JSON.stringify(doc, null, 4);
+    // Generate dynamic fields
+    var authors = generatePersons('author');
+    if (authors.length > 0) {
+        doc["author"] = authors;
+    }
+    var contributors = generatePersons('contributor');
+    if (contributors.length > 0) {
+        doc["contributor"] = contributors;
+    }
+    return doc;
+}
+
+// FIXME this hack to make expansion work for now
+// (expansion does not accept empty values)
+function deleteEmptyValues(doc) {
+  return JSON.parse(JSON.stringify(doc));
+}
+
+async function generateCodemeta() {
+    var inputForm = document.querySelector('#inputForm');
+    var codemetaText, codemetaTextToValidate, errorHTML;
+
+    if (inputForm.checkValidity()) {
+        var doc = buildDoc();
+        const expanded = await jsonld.expand(deleteEmptyValues(doc), {documentLoader: jsonldCustomLoader});
+        const compacted = await jsonld.compact(expanded, CODEMETA_CONTEXT_URL, {documentLoader: jsonldCustomLoader});
```

Comment:
do we actually need to expand here? can't the compaction work directly on `doc`?

Reply:
The compaction does not seem to work directly on `doc` because of nested values (which should have an additional `@list` property). Also, I felt that it was best practice to work on the compacted value, and especially to be consistent with the import later on, and maybe to help with validation (though I agree it looks weird as it is now).

Reply:
Ah, indeed. Eventually we should make the main code generate the expanded value directly, but let's keep this for a future PR.

Reply:
Agreed! Created https://github.com/codemeta/codemeta-generator/issues/29.